message_fetch: Convert to typed_endpoint.

kennethnrk opened the pull request:
Migrated `message_fetch` to `typed_endpoint`.

Removed the old `narrow_parameter` as we have shifted to the new `NarrowParameter` Pydantic Object.

This new object provides better error messages for data validation, hence changed the error messages in `test_message_fetch`. 

`num_before` and `num_after` are required parameters, added them to all the calls of `get_messages_backend`.


### narrow: Use NarrowParameter instead of dictionary.
Earlier we were using the type `OptionalNarrowListT` for all functions
that required "narrow" as a parameter.
This commit changes all the functions accepting a "narrow"
to use a list of the new `NarrowParameter`
instead of `OptionalNarrowListT` which is a list of dicts.

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -120,6 +123,11 @@ def get_messages_backend(
+    if narrow is not None and len(narrow) > 0:
+        narrow_dict = [x.model_dump() for x in narrow]
+    else:
+        narrow_dict = None
```
Why is this called `narrow_dict`? It's a list, not a dictionary. Maybe `narrow_parameter_list` would be an OK variable name. What is the type we get out of `model_dump` here; can we declare it explicitly?

`narrow_parameter_list: OptionalNarrowListT = ...` seems a lot more readable.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -3520,10 +3526,19 @@ def test_bad_narrow_type(self) -> None:
+            "narrow is not valid JSON",
+            "Invalid narrow[0]: Value error, dict or list required",
+            "Invalid narrow[0]: Value error, element is not a string pair",
+        ]
```
This is fragile, do a list of pairs of `(value, expected_error)` rather than two parallel lists.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -15,7 +15,6 @@
     Union,
 )
 
```
The commit message should have the explanations that you have in the PR description for the changes; this is important because we merge commits, not PRs.

That said, I think you should split this into a couple commits -- the first should have the changes to add `num_before/num_after` where it was previously not passed, since that's basically an independent bug fix that's required to do the migration.

And then you can have a second one that changes the APIs of `exercise_bad_narrow_operand` and its friends to accept a list of possible error messages ... before a final commit that does the conversion itself.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -3542,50 +3557,59 @@ def test_invalid_narrow_operand_in_dict(self) -> None:
     def exercise_bad_narrow_operand_using_dict_api(
-        self, operator: str, operands: Sequence[Any], error_msg: str
+        self, operator: str, operands: Sequence[Any], error_msgs: List[str]
     ) -> None:
```
This parallel lists API is bad; I think you might just want to define a dataclass or tuple format or something for these operand/error pairs.

kennethnrk commented:
I've split the commit into 3 commits.

For the multiple error messages I've used a `List` of `Dict` as it seemed to be the most readable option.

Addressed the other comment as well.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -3483,18 +3483,25 @@ def test_bad_int_params(self) -> None:
+        ]
         for idx, param in enumerate(int_params):
-            for type in bad_types:
+            for type_dict in bad_types:
```
The `type_dict/bad_types` names feel fairly confusing at this point. 

Maybe call these `invalid_parameters`, and `invalid_parameter["value"]/invalid_parameter["expected_error"]`? That seems a lot more readable.

PIG208 replied:
`Dict[str, Any]` is not as restrictive as a dataclass so it is less preferrable. We could have something like:

```python
@dataclass
class InvalidParam:
    value: object
    expected_error: str
```

This can also be reused for later occurrences of `bad_types` in other tests.

PIG208 commented on zerver/tests/test_message_fetch.py:
```diff
@@ -3541,90 +3548,128 @@ def test_invalid_narrow_operand_in_dict(self) -> None:
     # test bad narrow, this method uses the current dict API format
     def exercise_bad_narrow_operand_using_dict_api(
-        self, operator: str, operands: Sequence[Any], error_msg: str
+        self, operator: str, operands: Sequence[Dict[str, Any]]
```
We should try to avoid the use of `Dict[str, Any]` and switch to things like dataclasses.. This way the type checker always ensures that these helpers are used properly.

PIG208 commented on zerver/tests/test_message_fetch.py:
```diff
@@ -3541,90 +3548,128 @@ def test_invalid_narrow_operand_in_dict(self) -> None:
 
     def exercise_bad_narrow_operand(
-        self, operator: str, operands: Sequence[Any], error_msg: str
+        self, operator: str, operands: Sequence[Dict[str, Any]]
```
Same here.

PIG208 commented on zerver/views/message_fetch.py:
```diff
@@ -120,6 +123,11 @@ def get_messages_backend(
         raise JsonableError(_("The anchor can only be excluded at an end of the range"))
 
+    if narrow is not None and len(narrow) > 0:
+        narrow_parameter_list: OptionalNarrowListT = [x.model_dump() for x in narrow]
```
`OptionalNarrowListT` is `Optional[List[Dict[str, Any]]]`. Now that we have `NarrowParameter`, can we drop `OptionalNarrowListT` and just use `Optional[List[NarrowParameter]]` with all the helper functions?

PIG208 commented on zerver/views/message_fetch.py:
```diff
@@ -134,11 +142,11 @@ def get_messages_backend(
-        assert narrow is not None
-        if not is_spectator_compatible(narrow):
+        assert narrow_parameter_list is not None
+        if not is_spectator_compatible(narrow_parameter_list):
```
At least functions like `clean_narrow_for_web_public_api`, `is_web_public_narrow`, `is_spectator_compatible` and `fetch_messages` will be able to use `narrow` without the `.model_dump()` calls.

PIG208 commented on zerver/views/message_fetch.py:
```diff
@@ -281,13 +289,16 @@ def get_messages_backend(
+    msg_ids: Json[List[int]],
+    narrow: Json[List[NarrowParameter]],
 ) -> HttpResponse:
+    narrow_parameter_list: OptionalNarrowListT = [x.model_dump() for x in narrow]
```
Similar here with `OptionalNarrowListT`. The refactoring might be done in the same commit, or separately in a commit following this change.

kennethnrk commented:
@PIG208 Thanks for the review. I have addressed the comments regarding the `InvalidParam` in `test_message_fetch.py` by creating a dataclass. I have also made a new commit to change all the functions in `narrow.py` to use `NarrowParameter` instead of the earlier type alias.

timabbott commented:
Looks very nice, merged, thanks @kennethnrk! I think this was likely one of the most complex endpoints to migrate, so glad to have it cut over :)

PIG208 commented:
Thanks for the work @kennethnrk! Feel free to ping me to review `typed_endpoint`-related changes like this. I would like to help.
